warehouse/packaging/metadata_verification.py

```diff
@@ -41,6 +50,86 @@ def _verify_url_pypi(url: str, project_name: str, project_normalized_name: str)
     )
 
 
+def _get_url_content(url: rfc3986.api.URIReference, max_length_bytes: int) -> str:
+    """
+    Get at least `max_length_bytes` of the contents of a URL
+
+    Raises `requests.exception.RequestException` on any errors while trying to
+    access the URL.
+    """
+    headers = {"User-Agent": "warehouse/1.0 (PyPI URL verifier, https://pypi.org)"}
+    r = requests.get(
+        url.unsplit(),
+        allow_redirects=False,
+        timeout=5,
+        stream=True,
+        headers=headers,
+    )
+    r.raise_for_status()
+
+    content = next(r.iter_content(max_length_bytes))
+    return content
```

Comment:
Thinking out loud: should we explicitly `r.close()` before existing this method? As-is, I _think_ this will leave the connection open and dangling until the server hangs up, meaning that we could end up slowly exhausting the number of available outbound sockets.

Reply:
Good catch! This also applies to the new `urllib3` implementation, so we now call 
```python
    r.drain_conn()
    r.release_conn()
```
before returning

Reply:
Hmm, is there a way we can avoid `drain_conn`? My understanding is that `r.drain_conn(...)` will block until the entire response is read (and dropped), meaning that this will end up effectively reading the whole response instead of truncating after the first X bytes.

Per: https://urllib3.readthedocs.io/en/stable/reference/urllib3.response.html#urllib3.response.HTTPResponse.drain_conn

(If I got that right, then I _think_ we can avoid this by removing the call and doing nothing else, since we're already using a separate connection pool per request and there's no reason to release back to a pool that we don't reuse 🙂)

Reply:
I think that makes sense, but shouldn't we close the connection? The [docs](https://urllib3.readthedocs.io/en/stable/advanced-usage.html#streaming-and-i-o) mention that if you don't care about returning the connection to the pool, you can call `close()` to close the connection:

> You can call the [close()](https://urllib3.readthedocs.io/en/stable/reference/urllib3.response.html#urllib3.response.HTTPResponse.close) to close the connection, but this call doesn’t return the connection to the pool, throws away the unread data on the wire, and leaves the connection in an undefined protocol state. This is desirable if you prefer not reading data from the socket to re-using the HTTP connection.

Reply:
Yeah I recommend calling `HTTPResponse.close()` instead of `drain_conn()` if you don't plan on reusing the connection, that will close out the socket.

Reply:
fixed!